rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -865,6 +872,116 @@ extern "C"
         return RMW_RET_OK;
     }
 
+    rmw_waitset_t *
+    rmw_create_waitset(rmw_guard_conditions_t * fixed_guard_conditions, size_t max_conditions)
+    {
+      rmw_waitset_t * waitset = rmw_waitset_allocate();
+      GuardCondition * rtps_guard_cond = nullptr;
+      CustomWaitsetInfo * waitset_info = nullptr;
+
+      // From here onward, error results in unrolling in the goto fail block.
+      if (!waitset) {
+        RMW_SET_ERROR_MSG("failed to allocate waitset");
+        goto fail;
+      }
+      waitset->implementation_identifier = eprosima_fastrtps_identifier;
+      waitset->data = rmw_allocate(sizeof(CustomWaitsetInfo));
+      // This should default-construct the fields of CustomWaitsetInfo
+      waitset_info = static_cast<CustomWaitsetInfo *>(waitset->data);
+      RMW_TRY_PLACEMENT_NEW(waitset_info, waitset_info, goto fail, CustomWaitsetInfo);
+      if (!waitset->data || !waitset_info) {
+        RMW_SET_ERROR_MSG("failed to construct waitset info struct");
+        goto fail;
+      }
+
+      waitset->fixed_guard_conditions = fixed_guard_conditions;
+      if (fixed_guard_conditions) {
+        if (!fixed_guard_conditions->guard_conditions) {
+          RMW_SET_ERROR_MSG("Received invalid guard condition array");
+          goto fail;
+        }
+        // We also need to attach the fixed guard conditions to the waitset (and detach them in destroy)
+        for (size_t i = 0; i < fixed_guard_conditions->guard_condition_count; ++i) {
+          void * guard_cond = fixed_guard_conditions->guard_conditions[i];
+          if (!guard_cond) {
+            RMW_SET_ERROR_MSG("Received invalid guard condition");
+            goto fail;
+          }
+          rtps_guard_cond = static_cast<GuardCondition *>(guard_cond);
+          rtps_guard_cond->attachCondition(&waitset_info->condition_mutex, &waitset_info->condition);
+        }
+      }
+
+      return waitset;
+
+    fail:
+      if (waitset) {
+        if (waitset->data) {
+          if (waitset_info) {
```

Comment:
This check seems to be unnecessary since it is essentially the same as `waitset->data`.